### PR TITLE
Colorscheme1.1 Heatmap

### PR DESCRIPTION
**Pull request by lanzhenw**
As a user, I can configure the heatmap field and float field colorscale through either providing a named [colorscale](https://plotly.com/python/builtin-colorscales/) or manually defining a value color range list. The field will first look for field specific colorscale defined, and then check `defaultColorscale` of the session. 

Set colorscale for a field on UI:

https://github.com/voxel51/fiftyone/assets/17770824/631d5360-e59c-4386-8826-dc544eb6274e

Set colorscale with manual defined list:

https://github.com/voxel51/fiftyone/assets/17770824/9df73145-fcd9-4615-a2e0-244be40ac1dd

Set colorscale for a float type field to render the embeddings:

https://github.com/voxel51/fiftyone/assets/17770824/a1edfd83-cde2-4eb4-a7fe-40cbd2bbb963

Example schema:
```
        import fiftyone as fo
        import fiftyone.zoo as foz

        dataset=foz.load_zoo_dataset("quickstart")

        # Store a custom color scheme for a dataset
        dataset.app_config.color_scheme = fo.ColorScheme(
            color_by="field",
            color_pool=["#ff0000", "#00ff00", "#0000ff"],
            fields=[
                {
                    "path": "ground_truth",
                    "fieldColor": "#ff00ff",
                    "colorByAttribute": "label",
                    "valueColors": [{"value": "dog", "color": "yellow"}],
                    "maskTargetsColors": [{"intTarget": 2, "color": "#ff0000"}, {"intTarget": 12, "color": "#99ff00"}],
                }
            ],
            label_tags={
                "fieldColor": "#00ffff",
                "valueColors": [
                    {"value": "correct", "color": "#ff00ff"},
                    {"value": "mistake", "color": "#00ff00"},
                ]
            },
            colorscales = [

                {
                    "path": "float_field",
                    "list": [{"value": 0, "color": 'rgb(0, 0, 255)'}, {"value": 1, "color": 'rgb(0, 255, 255)'}],
                },
                {
                    "path": "heatmap",
                    "name": "hsv",
                },
            ],
            multicolor_keypoints=False,
            opacity=0.5,
            show_skeletons=True,
            default_mask_targets_colors=[{"intTarget": 1, "color": "#FEC0AA" }, {"intTarget": 2, "color": "#EC4E20"}],
            default_colorscale={"name": "sunset", "list": None}
        )

        session = fo.launch_app(dataset)
```

## What changes are proposed in this pull request?

This PR also includes changes in the backend session workflow of how colorscheme is handled. 
We added `id` field to the object (needs migration)

## How is this patch tested? If it is not, please explain why.
- I can define a default colorscale for the dataset. If a field does not have specific setting, default colorscale should be used. 
- I can define field level colorscale for float field and heatmap field. 
- Json viewer should capture instant changes. I can modify the schema in json viewer directly. 
- I can save a schema to the dataset. When loading a new session, it will use the dataset default setting. 

bugs:

- [ ] Regression: session.color_scheme = fo.Colorscheme(color_by="field",
            color_pool=["#ff0000", "#00ff00", "#0000ff"]) is not working.
- [ ] Initial load: rgb is not loaded until we open the color panel and change certain settings
- [ ] `datasetColorscheme` fragment does not reliably update when `save as default`

- ## Release Notes

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [x] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

File: app/packages/core/src/components/ColorModal/controls/ManualColorScaleList.tsx
Comment (benjaminpkane) [inline app/packages/core/src/components/ColorModal/controls/ManualColorScaleList.tsx:34]: small typo
```suggestion
type ManualColorScaleListProp = {
```

File: fiftyone/core/odm/dataset.py
Comment (benjaminpkane) [inline fiftyone/core/odm/dataset.py:282]: I think this can be removed, right?

**Review by benjaminpkane — COMMENTED** [review]
I defer to @manivoxel51 on approval. Code wise, this looks pretty complete! Any fixes could happen in separate PRs from my perspective

**Review by manivoxel51 — APPROVED** [review]
Nice work!💥 

Tested the app with heatmaps and didn't notice any blocking issues. I agree that we can crush any known bugs as a followup PR.

**Review by manivoxel51 — COMMENTED** [review]
@lanzhenw, @benjaminpkane e2e is failing - can we get those fixed before merging?

**Review by manivoxel51 — COMMENTED** [review]
Two more things before merging

- Please remove all console.logs 

Screen Shot 2023-11-17 at 4 00 15 PM

- nit: there are 00 in the list value input.

Screen Shot 2023-11-17 at 3 58 55 PM
e use single 0

**Comment by lanzhenw** [comment]
@manivoxel51 I fixed the e2e tests and the typing issue. 

_typing 000 => 0
typing 02 => 2
typing 0.00 => 0.00
typing 0 => 0_ 

https://github.com/voxel51/fiftyone/assets/17770824/de3c8e72-53dd-4b4b-a606-6e23e1199422

**Review by manivoxel51 — APPROVED** [review]
we'll investigate the failing test - locally passing - CI failing

**Pull request merged**
